More AES guards testing and some fixes

daverodgman opened the pull request:
## Description

Extend AES guards testing to cover `MBEDTLS_AES_USE_HARDWARE_ONLY` and `MBEDTLS_AES_ONLY_128_BIT_KEY_LENGTH`, which were added after the test; also `MBEDTLS_AESCE_C` and `MBEDTLS_AESNI_C` which were missed out.

Fix the issues uncovered by the extra testing.

## PR checklist

- [x] **changelog** not required
- [x] **backport** not required - not significant enough to backport
- [x] **tests** provided



## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

tom-cosgrove-arm commented:
I'm happy for the `aes.c` fixes to go in, but less happy about the increase in CI time for the tests. We really need an "EP" (for those that remember) CI run, that we can do locally and something like once a week on `development`

davidhorstmann-arm commented:
> nearly 10 times longer

3 extra nested loops of {set, unset} => 2^3 = 8 times longer would be expected, so it's the right order of magnitude.

gilles-peskine-arm commented:
On the CI, a single-threaded run of PR tests would probably take over 30 hours by now (it was about 20 hours a few versions ago, and since then we've added several driver-only builds and TLS 1.3 test runs). So +6 min is barely noticeable.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -3934,27 +3934,58 @@ component_build_tfm() {
+        echo ./scripts/config.py $h MBEDTLS_AES_USE_HARDWARE_ONLY
+        echo ./scripts/config.py $i MBEDTLS_AESNI_C
+        echo ./scripts/config.py $j MBEDTLS_AESCE_C
+        echo ./scripts/config.py $k MBEDTLS_AES_ONLY_128_BIT_KEY_LENGTH
```
If you (Tom, whoever) is looking to optimize, `config.py` is not particularly fast. Adding a mode that allows you set/unset a bunch of options in a single invocation might make a significant difference here.

The reason there's no such option is that `config.py` is backward-compatible with its ancestor `config.pl`. The perl code was faster, but much less flexible.

gilles-peskine-arm replied:
Or even faster: unset all of these options before the loop, then use `-D` on the compiler command line.

Also, you could invoke `clang` directly instead of running `make`: you're only building a single file, in a controlled environment.

daverodgman replied:
Yes, I was experimenting locally with using `sed` instead. That, plus skipping make and using `clang -fsyntax-only` gets it from 7m down to 1:52, which is worth having.

tom-cosgrove-arm replied:
Actually, if `config.py` is slow, just moving the loops that might-not-be-run further out from innermost would save time

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -3934,45 +3934,91 @@ component_build_tfm() {
+
+    echo "all: ${DEPS}" >> $MAKEFILE
+
+    NCPUS=$(lscpu -p|tail -n1|sed 's/,.*//')
```
We already have this calculation in `all_sh_nproc` with support for more OSes. But please respect the `MAKEFLAGS` setting.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -3934,45 +3934,91 @@ component_build_tfm() {
+    MAKEFILE=$(mktemp)
+    DEPS=""
+
+    for a in 0 1; do [[ $a == 0 ]] && A="" || A="-DMBEDTLS_AES_SETKEY_ENC_ALT"
```
I don't understand the point of the integer-valued variables. They're just making target names hard to read. Why not simply

```
for a in "" -DMBEDTLS_AES_SETKEY_ENC_ALT; do
…
echo "t$a$b…: ; clang $a $b …"
```

tom-cosgrove-arm commented on tests/scripts/all.sh:
```diff
@@ -3934,45 +3934,92 @@ component_build_tfm() {
+    rm ${MAKEFILE}
+}
+
+build_aes_variations_validate_combo() {
```
Since the function names in `all.sh` typically start with a significant word (`support_`, `component_`, etc) why not start the name with `validate_`?

tom-cosgrove-arm commented on tests/scripts/all.sh:
```diff
@@ -3946,9 +3946,24 @@ build_test_config_combos() {
+    # This ensures that we have any include paths, macro definitions, etc
+    # that may be applied by make.
+    # Add -fsyntax-only as we only want a syntax check and don't need to generate a file.
+    MAKE_CMD=$(make -B -n ${FILE} CC=clang CFLAGS="${WARNING_FLAGS} -fsyntax-only" | egrep "^clang")
```
Isn't this `COMPILE_CMD`? We might have got it from `make`, but...

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -3934,45 +3934,102 @@ component_build_tfm() {
+        TARGET="t"
+        CLANG_ARGS=""
+        for ((j = 0; j < ${LEN}; j++)); do
+            OPT=${OPTIONS[j]}
+            X=$(((i >> j) & 1))
+            [[ $X == 0 ]] && OPT="" || OPT="-D${OPT}"
```
Here and elsewhere, please use `if` rather than obscure conditionals. On the other hand, the multiple layers of option extraction could be simplified.
```suggestion
            OPT=
            if (((i >> j) & 1)); then
                OPT=-D${OPTIONS[j]}
            fi
```

gilles-peskine-arm reviewed (changes requested):
I think the code is correct, but it's hard to read.

At this level of complexity, it might be better to switch to Python. (I'm comfortable with bash, but most maintainers of Mbed TLS are more comfortable with Python.)

daverodgman commented on tests/scripts/all.sh:
```diff
@@ -3934,45 +3934,102 @@ component_build_tfm() {
+    # The most common issue is unused variables/functions, so ensure -Wunused is set.
+    WARNING_FLAGS="-Werror -Wall -Wextra -Wwrite-strings -Wpointer-arith -Wimplicit-fallthrough -Wshadow -Wvla -Wformat=2 -Wno-format-nonliteral -Wshadow -Wasm-operand-widths -Wunused"
+
+    # Extract the command generated by the Makefile to build the target file.
+    # This ensures that we have any include paths, macro definitions, etc
+    # that may be applied by make.
+    # Add -fsyntax-only as we only want a syntax check and don't need to generate a file.
+    COMPILE_CMD=$(make -B -n ${FILE} CC=clang CFLAGS="${WARNING_FLAGS} -fsyntax-only" | egrep "^clang")
```
This came about via discussion with @tom-cosgrove-arm , on the basis that it ensures we stay in sync with what Make actually does. I don't have a strong preference, but maybe being slightly fragile is better than potentially silently deviating from make behaviour? Tom WDYT?

gilles-peskine-arm replied:
How about a compromise that leverages the existing makefile? Let it set `LOCAL_CFLAGS` and use that.
```
include library/Makefile

t-%: aes.c; clang -Whard-coded-stuff $(LOCAL_FLAGS) -fsyntax-only aes.c
```

daverodgman replied:
Unfortunately including `library/Makefile` breaks (even from a Makefile containing only the line `include library/Makefile`). I couldn't find a simple way to make this work.

tom-cosgrove-arm replied:
> maybe being slightly fragile is better than potentially silently deviating from make behaviour?

Yes, that would be my preference, rather than building up a compiler invocation manually

gilles-peskine-arm replied:
> Unfortunately including library/Makefile breaks

Got it: https://github.com/Mbed-TLS/mbedtls/pull/6915#discussion_r1346244686

gilles-peskine-arm replied:
See https://github.com/gilles-peskine-arm/mbedtls/tree/more-aes-checks%2Binclude_Makefile for:

1. A fix for `include library/Makefile` not working.
2. A compromise between the fragile parsing and hard-coding things that should be in the makefile. My current code gets the official `LOCAL_FLAGS`, but doesn't build the dependencies of the normal rule to build the object file. I haven't found a way to do that last thing yet, except with some very akward changes to `library/Makefile`.

(Feel free to ignore the file refactoring, I only did that to make testing easier.)

daverodgman replied:
Thanks, have incorporated this into the latest iteration

daverodgman commented on tests/scripts/all.sh:
```diff
@@ -3934,45 +3934,102 @@ component_build_tfm() {
+    # This ensures that we have any include paths, macro definitions, etc
+    # that may be applied by make.
+    # Add -fsyntax-only as we only want a syntax check and don't need to generate a file.
+    COMPILE_CMD=$(make -B -n ${FILE} CC=clang CFLAGS="${WARNING_FLAGS} -fsyntax-only" | egrep "^clang")
```
That's odd. For me (Ubuntu 22.04 in a VM on M1 Pro), clang is faster (92 ms vs 135 ms).

daverodgman replied:
Overall for me, it's about 26s with gcc vs 18-20 with clang

daverodgman commented:
> At this level of complexity, it might be better to switch to Python

Probably true, but I think let's stick with bash for now and when `all.sh` migrates to Python, we can switch then.

tom-cosgrove-arm commented on library/aes.c:
```diff
@@ -352,11 +355,12 @@ static const uint32_t RT3[256] = { RT };
 #endif /* !MBEDTLS_AES_FEWER_TABLES */
 
-#endif /* !defined(MBEDTLS_AES_DECRYPT_ALT) || !defined(MBEDTLS_AES_SETKEY_DEC_ALT) */
+#endif \
```
Not entirely sure the

tom-cosgrove-arm commented on tests/scripts/all.sh:
```diff
@@ -3934,45 +3934,102 @@ component_build_tfm() {
+                opt=-D${options[j]}
+            fi
+            clang_args="${clang_args} ${opt}"
+            target="${target}${opt}"
```
can't these two go inside the `if`, and then you get rid of the `opt=(nothing)` at line 3981?

tom-cosgrove-arm commented on tests/scripts/all.sh:
```diff
@@ -3934,45 +3934,102 @@ component_build_tfm() {
+        for ((j = 0; j < ${len}; j++)); do
+            opt=
+            if (((i >> j) & 1)); then
+                opt=-D${options[j]}
```
I have a mild preference for `options[$j]`, i.e. having the sigil `$` when read the shell variable

tom-cosgrove-arm replied:
@daverodgman Does the rocket mean you are changing this to

```
opt=-D${options[$j]}
```

tom-cosgrove-arm commented on tests/scripts/all.sh:
```diff
@@ -3934,45 +3934,102 @@ component_build_tfm() {
+validate_aes_config_variations() {
+    if [[ "$1" == *"MBEDTLS_AES_USE_HARDWARE_ONLY"* ]]; then
+        if [[ "$1" == *"MBEDTLS_PADLOCK_C"* ]]; then
+            echo 1
```
might be better to `echo INVALID`

tom-cosgrove-arm replied:
@daverodgman Does the rocket mean you are `s/1/INVALID/` ?

tom-cosgrove-arm commented on tests/scripts/all.sh:
```diff
@@ -3934,45 +3934,102 @@ component_build_tfm() {
+    # This ensures that we have any include paths, macro definitions, etc
+    # that may be applied by make.
+    # Add -fsyntax-only as we only want a syntax check and don't need to generate a file.
+    compile_cmd=$(make -B -n ${file} CC=clang CFLAGS="${warning_flags} -fsyntax-only" | egrep "^clang")
```
I definitely prefer asking `make` what it would run. Yes, can be fragile, but no perfect solution here

tom-cosgrove-arm commented on library/aes.c:
```diff
@@ -84,8 +84,10 @@ static int aes_padlock_ace = -1;
+#if !defined(MBEDTLS_AES_ENCRYPT_ALT) || \
+    (!defined(MBEDTLS_AES_SETKEY_ENC_ALT) && (!defined(MBEDTLS_AES_USE_HARDWARE_ONLY) || \
+    !defined(MBEDTLS_AES_ROM_TABLES))) || \
+    (!defined(MBEDTLS_AES_SETKEY_DEC_ALT) && !defined(MBEDTLS_AES_USE_HARDWARE_ONLY))
```
this is pretty much unreadable

The only way of making it readable I found is to have too long a line, so no good suggestion here.

```
#if !defined(MBEDTLS_AES_ENCRYPT_ALT) || \                                                                                                                                   
    (!defined(MBEDTLS_AES_SETKEY_ENC_ALT) && (!defined(MBEDTLS_AES_USE_HARDWARE_ONLY) || !defined(MBEDTLS_AES_ROM_TABLES))) || \                                             
    (!defined(MBEDTLS_AES_SETKEY_DEC_ALT) && !defined(MBEDTLS_AES_USE_HARDWARE_ONLY))
```

daverodgman replied:
I considered simply marking these variables as "maybe unused" with `__attribute__((unused))`, and then all the variable declarations can just be unconditionally present. They are static so there's no impact on code size.

The places where they are used remain awkward though.

I was considering a separate PR for this, but I could bring it into this one if you prefer?

tom-cosgrove-arm replied:
I think marking them as "maybe unused" is even worse - the idea of this PR is (to me) to get what's needed correctly specified. Just saying "here's some data, please, compiler, remove it if it's not needed and don't tell me about it" seems... wrong.

I was wondering about building up a `#define NEED_F_SBOX` or some such, but it doesn't really help the readability.

I guess we're stuck with this one

daverodgman replied:
I also explored using `NEED_XYZ`, but agreed, it doesn't help very much.

Actually I do think leaving it to the compiler helps - it greatly simplifies the definitions. I don't think it can introduce any risk?

daverodgman replied:
I've pushed it anyway; see what you think. On balance I think it is definitely an improvement - why manually maintain complex `#if ...` blocks when the compiler can do a more reliable job with no complexity in the codebase?

gilles-peskine-arm replied:
In general I prefer to let the compiler figure out details like eliminating variables that are not used in some configurations, eliminating checks that are no-ops in some configurations, etc. The problem is to disable the warnings on all the compilers we care about.

tom-cosgrove-arm commented on library/aes.c:
```diff
@@ -352,11 +355,12 @@ static const uint32_t RT3[256] = { RT };
 #endif /* !MBEDTLS_AES_FEWER_TABLES */
 
-#endif /* !defined(MBEDTLS_AES_DECRYPT_ALT) || !defined(MBEDTLS_AES_SETKEY_DEC_ALT) */
+#endif \
```
(aside) it seems a bit weird having a line-continuation character to add a comment that will be removed at an earlier stage of processing anyway

gilles-peskine-arm replied:
That needs to be submitted as a request/patch to uncrustify.

tom-cosgrove-arm commented on library/common.h:
```diff
@@ -334,4 +334,25 @@ static inline void mbedtls_xor_no_simd(unsigned char *r,
+#define MBEDTLS_MAYBE_UNUSED __attribute__((unused))
+#endif
+#endif
+#if !defined(MBEDTLS_MAYBE_UNUSED) && defined(_MSC_VER)
```
IIRC when clang is used as the compiler driver by Visual Studio, it doesn't define `__GNUC__` and does define `_MSC_VER`, but only understands the GCC-style attributes, so I think we need

```
#if !defined(MBEDTLS_MAYBE_UNUSED) && (defined(__GNUC__) || defined(__clang__)
```

I did a quick'n'dirty install of this in a VM for testing some weeks ago - I'll see if I can dig it up and test again

daverodgman replied:
I suspect a lot of the codebase will get this wrong, but I'll fix this instance.

gilles-peskine-arm replied:
Would moving the `defined(__has_attribute)` up take care of that case? (Even with that, we do want the `__GNUC__` case, at least, for older GCC versions that predate `__has_attribute`.)

gilles-peskine-arm reviewed (changes requested):
The changes up to 4b779bef9eec31859e16c11df507235d86aa8e28 look good to me, but they're incomplete. Please resolve https://github.com/Mbed-TLS/mbedtls/pull/8283#discussion_r1344319617 and https://github.com/Mbed-TLS/mbedtls/pull/8283#discussion_r1344373707.

daverodgman commented:
> The changes up to [4b779be](https://github.com/Mbed-TLS/mbedtls/commit/4b779bef9eec31859e16c11df507235d86aa8e28) look good to me, but they're incomplete. Please resolve [#8283 (comment)](https://github.com/Mbed-TLS/mbedtls/pull/8283#discussion_r1344319617) and [#8283 (comment)](https://github.com/Mbed-TLS/mbedtls/pull/8283#discussion_r1344373707).

I think both of these are covered. The first via your make fix (thanks), the second via b1107aeee1fcb3afced5dea75171e8ec334c0c5d 

I have moved the define for `has_attribute` as we get a slight simplification from this, although still need all the cases. I've added some indentation for clarity.

gilles-peskine-arm reviewed (changes requested):
2457bcd26c2c8b803dce6318432b2a4f31ffa0bf looks good to me except that https://github.com/Mbed-TLS/mbedtls/pull/8283#discussion_r1344373707 is not addressed. Using an empty output to indicate success in bash is neither idiomatic nor efficient. Please change the validation function interface to returning 0 for valid and ≠0 for invalid, with no output on stdout.

gilles-peskine-arm reviewed (commented):
41bc798d7cb258871ab13f3122b5971a4ce64be6 looks good to me.

gilles-peskine-arm reviewed (approved):
41bc798d7cb258871ab13f3122b5971a4ce64be6 looks good to me.

tom-cosgrove-arm commented on tests/scripts/all.sh:
```diff
@@ -3919,45 +3919,107 @@ component_build_tfm() {
+
+    echo "build_test_config_combos: ${deps}" >> ${makefile}
+
+    # execute all of the commands via Make (probably in parallel)
```
> probably in parallel

What am I missing that's going to cause this? I don't see `-j`...

https://www.gnu.org/software/make/manual/html_node/Parallel.html:

> Normally, make will execute only one recipe at a time, waiting for it to finish before executing the next. However, the `‘-j’` or `‘--jobs’` option tells make to execute many recipes simultaneously.

daverodgman replied:
AIUI that is set (I guess via Makeflags) elsewhere by the CI

daverodgman replied:
Actually, it's only -j2 in the CI, but default is -j in all.sh: https://github.com/Mbed-TLS/mbedtls-test/blob/02a1d41b66f8675b77b71725dc137a0ef1e06ca2/vars/environ.groovy#L28C4-L28C4

tom-cosgrove-arm reviewed (approved):
LGTM
